File: guess.nim
{.experimental: "strictEffects".}
import std/random
import std/strformat
import std/strutils

type
  Game = object
    answer: int
    done: bool
    guesses: int
    high: int

# Treat as debug info only, for logging purposes.
var errCount = 0

func askGuess(high: int): int {.
    raises: [IOError, ValueError],
    tags: [ReadIOEffect, WriteIOEffect],
  .} =
  {.cast(noSideEffect).}:
    stdout.write &"Guess a number between 1 and {high}: "
    stdin.readLine.parseInt

func askGuessMulti(high: int): int {.
    raises: [IOError],
    tags: [ReadIOEffect, WriteIOEffect],
  .} =
  while true:
    try:
      return askGuess(high)
    except ValueError:
      {.cast(noSideEffect).}:
        echo "I didn't understand"
        errCount += 1

proc pickAnswer(high: int): int {.tags: [Rand].} =
  rand(high - 1) + 1

# proc pickAnswer(r: var Rand, high: int): int {.noSideEffect.} =
func pickAnswer(r: var Rand, high: int): int =
  r.rand(high - 1) + 1

proc report(game: Game, guess: int) {.tags: [WriteIOEffect]} =
  let description =
    if guess < game.answer: "too low"
    elif guess > game.answer: "too high"
    else: "the answer!"
  echo &"{guess} is {description}"

func update(game: Game, guess: int): Game =
  result = game
  if guess == result.answer:
    result.done = true;
  result.guesses += 1

proc play(game: var Game) {.tags: [ReadIOEffect, WriteIOEffect].} =
  while not game.done:
    let guess = askGuessMulti(game.high)
    game.report(guess)
    game = game.update(guess)

# proc main() {.tags: [Rand, ReadIOEffect, WriteIOEffect].} =
proc main() {.tags: [ReadIOEffect, WriteIOEffect].} =
  var r = initRand()
  let
    high = 100
    # answer = pickAnswer(high)
    answer = r.pickAnswer(high)
  var game = Game(answer: answer, done: false, guesses: 0, high: high)
  game.play
  echo &"Finished in {game.guesses} guesses"
  stderr.writeLine &"Total input errors: {errCount}"

main()

# func sneaky() =
#   var r = initRand()
#   discard pickAnswer(r, 100)

File: Guess.cs
int errCount = 0;

Main();

int AskGuess(int high) {
    Console.Write($"Guess a number between 1 and {high}: ");
    var text = Console.ReadLine();
    if (text == null) {
        throw new EndOfStreamException();
    }
    return int.Parse(text);
}

int AskGuessMulti(int high) {
    while (true) {
        try {
            return AskGuess(high);
        } catch (FormatException) {
            Console.WriteLine("I didn't understand");
            errCount += 1;
        }
    }
}

void PickAnswer(Random random, int high, out int answer) {
    // Inherently mutable random.
    answer = random.Next(high) + 1;
}

void Play(ref Game game) {
    while (!game.Done) {
        var guess = AskGuessMulti(game.High);
        Report(game, guess);
        game = Update(game, guess);
    }
}

void Report(Game game, int guess) {
    var description = guess switch {
        _ when guess < game.Answer => "too low",
        _ when guess > game.Answer => "too high",
        _ => "the answer!",
    };
    Console.WriteLine($"{guess} is {description}");
}

Game Update(Game game, int guess) {
    if (guess == game.Answer) {
        game.Done = true;
    }
    game.Guesses += 1;
    return game;
}

void Main() {
    var high = 100;
    int answer;
    PickAnswer(new Random(), high, out answer);
    var game = new Game { Answer = answer, High = high };
    // game = Play(game);
    Play(ref game);
    Console.WriteLine($"Finished in {game.Guesses} guesses");
    Console.WriteLine($"Total input errors: {errCount}");
}

record struct Game {
    public int Answer;
    public bool Done = false;
    public int Guesses = 0;
    public int High;
}
